src/Pim/Component/Catalog/spec/Completeness/CompletenessCalculatorSpec.php

```diff
@@ -0,0 +1,131 @@
+    {
+        $this->shouldHaveType(CompletenessCalculator::class);
+    }
+
+    function it_is_a_completeness_calculator()
+    {
+        $this->shouldImplement(CompletenessCalculatorInterface::class);
+    }
+
+    function it_does_not_calculates_completeness_for_a_product_without_family(ProductInterface $product)
+    {
+        $product->getFamily()->willReturn(null);
+
+        $completenesses = $this->calculate($product);
+        $completenesses->shouldBe([]);
+    }
+
+    function it_calculates_a_product_completeness(
+        $productValueFactory,
+        $channelRepository,
+        $localeRepository,
+        $productValueCompleteChecker,
+        ProductInterface $product,
+        FamilyInterface $family,
+        AttributeRequirementInterface $requirement,
+        ChannelInterface $channel,
+        ArrayCollection $locales,
+        \ArrayIterator $localesIterator,
+        LocaleInterface $locale,
+        AttributeInterface $attribute,
+        ProductValueCollectionInterface $requiredProductValues,
+        ProductValueCollectionInterface $actualProductValues,
+        ProductValueInterface $requiredProductValue
+    ) {
+        $product->getFamily()->willReturn($family);
+        $family->getAttributeRequirements()->willReturn([$requirement]);
+
+        $requirement->getChannel()->willReturn($channel);
+        $channel->getLocales()->willReturn($locales);
+
+        $locales->getIterator()->willReturn($localesIterator);
+        $localesIterator->rewind()->shouldBeCalled();
+        $localesIterator->valid()->willReturn(true, false, false);
+        $localesIterator->current()->willReturn($locale);
+        $localesIterator->next()->shouldBeCalled();
+
+        $requirement->isRequired()->willReturn(true);
+        $requirement->getChannelCode()->willReturn('channel_code');
+        $channel->getCode()->willReturn('channel_code');
+        $locale->getCode()->willReturn('locale_code');
+
+        $requirement->getAttribute()->willReturn($attribute);
+        $attribute->isLocaleSpecific()->willReturn(false);
+        $attribute->hasLocaleSpecific($locale)->shouldNotBeCalled();
+
+        $attribute->isScopable()->willReturn(false);
+        $attribute->isLocalizable()->willReturn(false);
+        $productValueFactory->create($attribute, null, null, null)->willReturn($requiredProductValue);
+        $channelRepository->findOneByIdentifier('channel_code')->willReturn($channel);
+        $localeRepository->findOneByIdentifier('locale_code')->willReturn($locale);
+
+        $product->getValues()->willReturn($actualProductValues);
+
+        $attribute->getCode()->willReturn('attribute_code');
+        $requiredProductValue->getScope()->willReturn('channel_code');
+        $requiredProductValue->getLocale()->willReturn('locale_code');
+        $requiredProductValues->getByCodes('attribute_code', 'channel_code', 'locale_code')->willReturn(null);
+        $productValueCompleteChecker->isComplete()->shouldNotBeCalled();
```

Comment:
hmmm maybe you should have another test (and rename this one)
with one or two required attributes?
what do you think?